File: programs/oilship/src/events.rs
//! Anchor events emitted by the OILSHIP program.

use anchor_lang::prelude::*;

#[event]
pub struct ProgramInitialized {
    pub admin: Pubkey,
    pub treasury: Pubkey,
    pub wreck_fund: Pubkey,
    pub toll_bps: u16,
    pub timestamp: i64,
}

#[event]
pub struct BridgeRegistered {
    pub bridge: Pubkey,
    pub symbol: String,
    pub operator: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct RiskUpdated {
    pub bridge: Pubkey,
    pub previous_score: u8,
    pub new_score: u8,
    pub previous_tier: u8,
    pub new_tier: u8,
    pub slot: u64,
    pub timestamp: i64,
}

#[event]
pub struct PolicyOpened {
    pub policy: Pubkey,
    pub beneficiary: Pubkey,
    pub bridge: Pubkey,
    pub cargo: u64,
    pub toll_paid: u64,
    pub class: u8,
    pub risk_at_open: u8,
    pub mature_slot: u64,
    pub expires_slot: u64,
    pub timestamp: i64,
}

#[event]
pub struct PolicySettled {
    pub policy: Pubkey,
    pub beneficiary: Pubkey,
    pub bridge: Pubkey,
    pub cargo: u64,
    pub timestamp: i64,
}

#[event]
pub struct WreckPayout {
    pub policy: Pubkey,
    pub beneficiary: Pubkey,
    pub bridge: Pubkey,
    pub principal_paid: u64,
    pub fund_balance_after: u64,
    pub timestamp: i64,
}

#[event]
pub struct BridgeQuarantined {
    pub bridge: Pubkey,
    pub final_score: u8,
    pub open_policies_at_quarantine: u32,
    pub coverage_at_quarantine: u64,
    pub timestamp: i64,
}

#[event]
pub struct QuarantineLifted {
    pub bridge: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct WreckFundDeposit {
    pub depositor: Pubkey,
    pub amount: u64,
    pub new_balance: u64,
    pub timestamp: i64,
}

#[event]
pub struct ConvoyOpened {
    pub convoy: Pubkey,
    pub bridge: Pubkey,
    pub opens_at_slot: u64,
    pub closes_at_slot: u64,
    pub timestamp: i64,
}

#[event]
pub struct PausedToggled {
    pub paused: bool,
    pub admin: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AdminTransferred {
    pub previous_admin: Pubkey,
    pub new_admin: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct ConfigUpdated {
    pub field: String,
    pub timestamp: i64,
}

#[event]
pub struct ThroughputThrottled {
    pub bridge: Pubkey,
    pub slot: u64,
    pub policies_in_slot: u32,
}

pub fn now_ts() -> Result<i64> {
    Ok(Clock::get()?.unix_timestamp)
}

File: sdk/src/events.ts
// Event subscription helpers for the OILSHIP SDK.

import { Bridge, Policy, Pubkey, Tier } from "./types.js";

export interface OilshipEvent {
  kind: EventKind;
  payload: Record<string, unknown>;
  slot: bigint;
  signature: string;
}

export type EventKind =
  | "ProgramInitialized"
  | "BridgeRegistered"
  | "RiskUpdated"
  | "PolicyOpened"
  | "PolicySettled"
  | "WreckPayout"
  | "BridgeQuarantined"
  | "QuarantineLifted"
  | "WreckFundDeposit"
  | "ConvoyOpened"
  | "PausedToggled"
  | "AdminTransferred"
  | "ConfigUpdated"
  | "ThroughputThrottled";

export interface RiskUpdatedPayload {
  bridge: Pubkey;
  previousScore: number;
  newScore: number;
  previousTier: Tier;
  newTier: Tier;
  slot: bigint;
}

export interface PolicyOpenedPayload {
  policy: Pubkey;
  beneficiary: Pubkey;
  bridge: Pubkey;
  cargo: bigint;
  tollPaid: bigint;
}

export interface WreckPayoutPayload {
  policy: Pubkey;
  beneficiary: Pubkey;
  bridge: Pubkey;
  principalPaid: bigint;
  fundBalanceAfter: bigint;
}

export class EventEmitter<T> {
  private readonly listeners: Set<(value: T) => void> = new Set();

  on(handler: (value: T) => void): () => void {
    this.listeners.add(handler);
    return () => this.listeners.delete(handler);
  }

  emit(value: T): void {
    for (const h of this.listeners) {
      try {
        h(value);
      } catch (err) {
        console.error("oilship event handler error:", err);
      }
    }
  }

  size(): number {
    return this.listeners.size;
  }
